File: Correlation.m
clc
clear all;
x=[3 4 5 6 2];
y=[4 5 6];
m=length(x);
n=length(y);
Lag=-(n-1):(m-1);
L=m+n-1;
znum=L-m;
Z=zeros(1,znum);
Xnew=[Z x Z];
for i=1:L
    A=zeros(1,L+znum);
    for k=1:n
        p=n-k;
    A(1,i+p)=y(1,p+1);
    end
    CRs=A.*Xnew;
    CRR(i)=sum(CRs);
end
Cross_correlation=[Lag;CRR]
subplot(3,1,1)
n1=1:m;
stem(n1,x,'filled','blue',linewidth=2)
title('x[n]')
xlabel('Sample Index')
text(n1, x, num2str(x', '%d\n'), 'VerticalAlignment', 'bottom', 'HorizontalAlignment', 'left')
subplot(3,1,2)
n2=1:n;
stem(n2,y,'filled','red',linewidth=2)
title('y[n]')
xlabel('Sample Index')
text(n2, y, num2str(y', '%d\n'), 'VerticalAlignment', 'bottom', 'HorizontalAlignment', 'left')
subplot(3,1,3)
n3=-(n-1):(m-1);
stem(n3,CRR,'filled','green',linewidth=2)
title('Correlation, r[n]')
xlabel('Lag Index')
text(n3, CRR, num2str(CRR', '%d\n'), 'VerticalAlignment', 'bottom', 'HorizontalAlignment', 'left')

File: Correlation.py
import numpy as np
import matplotlib.pyplot as plt

x = np.array([3, 4, 5, 6, 2])
y = np.array([4, 5, 6])
m = len(x)
n = len(y)
Lag = np.arange(-(n-1), m)
L = m + n - 1
znum = L - m
Z = np.zeros(znum)
Xnew = np.concatenate((Z, x, Z))
CRR = np.zeros(L)

for i in range(L):
    A = np.zeros(L + znum)
    for k in range(n):
        p = n - k - 1
        A[i + p] = y[p]
    CRs = A * Xnew
    CRR[i] = np.sum(CRs)

Cross_correlation = np.vstack((Lag, CRR))
print(Cross_correlation)
plt.subplot(3, 1, 1)
n1 = np.arange(1, m+1)
plt.stem(n1, x, 'b', use_line_collection=True)
plt.title('x[n]')
plt.xlabel('Sample Index')
for i, j in zip(n1, x):
    plt.text(i, j, str(j), va='bottom', ha='left')

plt.subplot(3, 1, 2)
n2 = np.arange(1, n+1)
plt.stem(n2, y, 'r', use_line_collection=True)
plt.title('y[n]')
plt.xlabel('Sample Index')
for i, j in zip(n2, y):
    plt.text(i, j, str(j), va='bottom', ha='left')

plt.subplot(3, 1, 3)
n3 = np.arange(-(n-1), m)
plt.stem(n3, CRR, 'g', use_line_collection=True)
plt.title('Correlation, r[n]')
plt.xlabel('Lag Index')
for i, j in zip(n3, CRR):
    plt.text(i, j, str(j), va='bottom', ha='left')

plt.tight_layout()
plt.show()

# r = np.correlate(x, y, 'full')
# print(r)
